cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -62,6 +62,57 @@ const FunctionDoc list_value_length_doc{
      "Null values emit a null in the output."),
     {"lists"}};
 
+template <typename Type, typename IndexType>
+Status ListElement(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+  using ListArrayType = typename TypeTraits<Type>::ArrayType;
+  using IndexScalarType = typename TypeTraits<IndexType>::ScalarType;
+  ListArrayType list_array(batch[0].array());
+  const auto& index_scalar = batch[1].scalar_as<IndexScalarType>();
+  auto index = index_scalar.value;
+  if (ARROW_PREDICT_FALSE(index < 0)) {
+    return Status::Invalid("Index ", index,
+                           " is out of bounds: should be greater than or equal to 0");
+  }
+  std::unique_ptr<ArrayBuilder> builder;
+  RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), list_array.value_type(), &builder));
+  for (int i = 0; i < list_array.length(); ++i) {
+    if (list_array.IsNull(i)) {
+      RETURN_NOT_OK(builder->AppendNull());
+      continue;
+    }
+    std::shared_ptr<arrow::Array> value_array = list_array.value_slice(i);
+    auto len = value_array->length();
+    if (ARROW_PREDICT_FALSE(index > len)) {
+      return Status::Invalid("Index ", index, " is out of bounds: should be in [0, ", len,
+                             "]");
+    }
+    ARROW_ASSIGN_OR_RAISE(auto scalar, (value_array->GetScalar(index)));
+    RETURN_NOT_OK(builder->AppendScalar(*scalar));
+  }
+  ARROW_ASSIGN_OR_RAISE(auto result, builder->Finish());
+  out->value = result->data();
+  return Status::OK();
+}
+
+template <typename InListType, typename IndexType>
+void AddListElement(ScalarFunction* func) {
```

Comment:
nit: AddListElementKernel?

Reply:
@lidavidm There is other function kernel that follows the same naming I did here ... I think is better to have homogeneous naming conventions even for these implementation details. See `AddListCast`

Reply:
Hmm, I'd say nearly all kernels follow the `AddFooKernel` pattern, looking at (for instance) scalar_if_else, or the aggregates, or the vector kernels. But casts, compare, and string kernels don't follow the pattern.

Reply:
I think these naming inconsistencies should be fixed at some point (maybe open a JIRA) bc as Arrow project grows it will be beneficial. Basically, this falls in the "Coding guidelines for Arrow" folder :)